### PR TITLE
[checkpoint] Log native DCP async save durations

### PR DESCRIPTION
**Pull request by ivy-zhou**
Summary:
Record native DCP asynchronous save latency from entry into dcp_save until the returned save future completes. Capture one monotonic timestamp at submission and attach a completion callback that subtracts it from the completion timestamp and emits the structured metric.

The metric measures the save future lifetime, not filesystem-only write time. The existing wait path remains responsible for surfacing save failures.

Test Plan:
env TORCH_DISABLE_ADDR2LINE=1 PYTHONPATH=/data/users/ivyzhou/torchtitan:/tmp/torchtitan_additional_packages_3e4a.nRk7S7/checkpointing:/tmp/torchtitan_additional_packages_3e4a.nRk7S7/ao:/tmp/torchtitan_additional_packages_3e4a.nRk7S7/torchtitan_runtime_deps /home/ivyzhou/local/torchtitan-venv/bin/python -m pytest tests/unit_tests/observability/test_structured_logging.py tests/unit_tests/test_checkpoint.py -q

138 passed.

Changed-file pre-commit hooks passed, including ufmt, flake8, pydoclint, codespell, and Pyrefly.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/torchtitan/pull/4279).
* #4457
* #4277
* __->__ #4279

### CODE REVIEW

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/components/checkpointer/dcp.py [hunk 1]
```diff
@@ -487,6 +490,18 @@ def _save(self, curr_step: int, last_step: bool = False) -> bool:
...
+            self.save_future.add_done_callback(
+                lambda _: sl.log_trace_scalar(
+                    {
+                        "train.checkpoint_write.native_dcp.execute.async_total.latency_ms": (
+                            time.monotonic() - async_save_started_at
+                        )
+                        * 1000
+                    }
+                )
+            )
```
Comment (fegin) [inline torchtitan/components/checkpointer/dcp.py:504]: This is only the staging cost, right?

**Review by fegin — APPROVED** [review]

File: torchtitan/components/checkpointer/dcp.py (hunk 1 above)
Comment (ivy-zhou) [reply torchtitan/components/checkpointer/dcp.py:504]: It should track the full save future no? I'm a bit confused why this is only the staging?
